### DISCUSSION TITLE
How to parse expression with an operator followed by two parameters

### DISCUSSION THREAD

**Discussion by DXie123** [petitparser/dart-petitparser#69] [Q&A] [answered]
I want to parse LaTex code like `\frac{1}{2}` (should be parsed to `1/2`).

After read the tutorial on README, I found support for format like `{parameter}operator{parameter}`, but I didn't find a way to parse code like `\frac{1}{2}`, which is `operator{parameter}{parameter}`

I want to know if there is an appropriate way to handle this problem.

**Comment by renggli** [comment]
Please consider asking general questions like this on https://stackoverflow.com/questions/tagged/dart+petitparser

That said, the following code parses your example:

```dart
final arg = char('{') & digit() & char('}');
final frac = string('\\frac') & arg.star();
```

**Comment by DXie123** [comment]
Thanks for your answer

I do tried some similar ways to parse it: 

```
final frac1 = string('\\frac{') & any().starLazy('}{') & char('}');  
final frac2 = string('\\frac{') & any().starGreedy('}{') & char('}');  
```

But none of them work in a more complex situation like `\frac{\frac{1}{2}}{3}+\frac{1}{2}`(should be (1/2)/3+1/2). Because there may have all kinds of code in parenthesis, I have to use `any()` in my project, but it always has some parenthesis mathcing problem.

**Comment by renggli** [comment, marked as answer]
To deal with nested grammars you need to recursively call yourself:

```dart
final frac = undefined();
final expr = frac | digit();
final arg = char('{') & expr & char('}');
frac.set(string('\\frac') & arg.star());

print(frac.parse('\\frac{\\frac{1}{2}}{3}')); 
// ==> Success[1:22]: [\frac, [[{, [\frac, [[{, 1, }], [{, 2, }]]], }], [{, 3, }]]]
```

This kind of recursive code is easier to write when using a [GrammarDefinition](https://pub.dev/documentation/petitparser/latest/petitparser/GrammarDefinition-class.html).
